--- mandos-monitor_ip	2020-05-15 14:06:46.003333507 -0400
+++ mandos-monitor	2020-04-08 14:46:32.000000000 -0400
@@ -207,12 +207,12 @@ class MandosClientWidget(urwid.FlowWidge
         log.info("Client %s received its secret",
                  self.properties["Name"])
 
-    def need_approval(self, timeout, default):
+    def need_approval(self, timeout, default, address):
         if not default:
-            message = "Client %s needs approval within %f seconds"
+            message = "Client %s from %s needs approval within %f seconds"
         else:
-            message = "Client %s will get its secret in %f seconds"
-        log.info(message, self.properties["Name"], timeout/1000)
+            message = "Client %s from %s will get its secret in %f seconds"
+        log.info(message, self.properties["Name"], address, timeout/1000)
 
     def rejected(self, reason):
         log.info("Client %s was rejected; reason: %s",

